projects/user_publication/gscholar.py

```diff
@@ -120,33 +124,83 @@ def get_authors(self, pub):
         Returns:
             list: list of authors in ["firstname lastname",] format
         """
-        bibt = pub['bibtex']
-        bibt = utils.decode_unicode_text(bibt)
-        match = re.search(r'author\s*=\s*\{([\w\s-,]+)\}', bibt)
-        if match:
-            authors_match = match.groups()[0]
-        else:
-            # this mostly never happens - if it does - log it
-            logger.info(f"cannot parse bibtex, so ignoring - {bibt}")
-            return
-        authors = authors_match.split(" and ")
+        authors = utils.decode_unicode_text(pub['bib']['author'])
+        # substitute all charachter from text except
+        # ',' , \w word charachter, \s whitespace charachter
+        # - any other charachter
+        # inspired from django.utils.text.slugify
+        authors = re.sub(r"[^\,\w\s-]", "", authors)
+        authors = authors.split(" and ")
         parsed_authors = [utils.parse_author(author) for author in authors]
         return parsed_authors
 
     def get_pub_model(self, pub):
-        return Publication(
+        entry_type = []
+        # bibtex ENTRYTYPE field
+        if 'ENTRYTYPE' in pub['bib']:
+            entry_type.append(pub['bib']['ENTRYTPE'])
+        if 'pub_type' in pub['bib']:
+            entry_type.append(pub['bib']['pub_type'])
+        forum = ''
+        if 'booktitle' in pub['bib']:
+            forum = pub['bib']['booktitle']
+        if 'journal' in pub['bib']:
+            forum = pub['bib']['journal']
+        pub_type = utils.get_pub_type(entry_type, forum)
+        pub_model = Publication(
             title=utils.decode_unicode_text(pub["bib"]["title"]),
             year=pub["bib"]["pub_year"],
             author=" and ".join(self.get_authors(pub)),
             entry_created_date=datetime.date.today(),
-            publication_type=utils.get_pub_type(pub["bib"]["ENTRYTYPE"]),
-            bibtex_source="{}",
+            publication_type=pub_type,
+            bibtex_source=pub['bib'],
             added_by_username="admin",
-            forum=pub["bib"]["booktitle"],
-            link=pub["pub_url"],
-            source="gscholar",
+            forum=forum,
+            link=pub.get("pub_url", ''),
+            source=Publication.G_SCHOLAR,

```

Comment:
What exactly is source for? I thought publications can have multiple sources.

Reply:
Where the publication first sources from -user_reported etc.,

Reply:
Yes, though as I mentioned, we only have one publication row per publication? So there will usually be multiple sources per publication

Reply:
Yes. I am making a change as mentioned by Adam to have PublicationSources as a model to accommodate the information about the name of the source the publication found in and the citation count.